src/data/simple_csr_source.cu

```diff
@@ -34,80 +38,165 @@ __global__ void CountValidKernel(common::Span<Columnar const> columns,
     return;
   }
   RBitField8 const mask = columns[bid].valid;
-  for (auto r = tid; r < n_rows; r += kBlockThreads) {
-    if (mask.Data() == nullptr || mask.Check(r)) {
-      atomicAdd(reinterpret_cast<BitFieldAtomicType*>(&offsets[r+1]),
-                static_cast<BitFieldAtomicType>(1));
+  if (std::isnan(missing)) {
+    // no missing value is specified
+    for (auto r = tid; r < n_rows; r += kBlockThreads) {
+      if (mask.Data() == nullptr || mask.Check(r)) {
+        atomicAdd(reinterpret_cast<BitFieldAtomicType*>(&offsets[r+1]),
+                  static_cast<BitFieldAtomicType>(1));
+      }
+    }
+  } else {
+    for (auto r = tid; r < n_rows; r += kBlockThreads) {
+      if (!common::CloseTo(columns[bid].data[tid], missing)) {
```

Comment:
Can you explain a bit more the need for CloseTo?

Reply:
Also, if I select NaN as my missing value, will this algorithm remove NaNs from the CSR?

Reply:
If NaN is selected, the algorithm will use the validity mask.  I can see the problem here.  Will try to address it.

Reply:
@RAMitchell Just a equality test for floating point value.  But in here we need to do it for both floating point and integer.  So I created a `constexpr` for dispatching the equality test based on input type.

Reply:
@RAMitchell 

> Also, if I select NaN as my missing value, will this algorithm remove NaNs from the CSR?

I added some comments above `SimpleCSRSource::CopyFrom` definition, hope that can clear things up.  When we add missing value support for other data types, those comment can be used as a reference and eventually should be put into user document.